File: done/kyu-5/scramblies.py
# https://www.codewars.com/kata/scramblies/train/python

def scramble(s1, s2):
    def get_map(s):
        s = sorted(s)
        su = set(s)
        return {c: s.count(c) for c in su}
    s1u = get_map(s1)
    s2u = get_map(s2)
    for char in s2u.keys():
        if char not in s1u.keys(): return False
        if s2u[char] > s1u[char]: return False
    return True

# tests

import unittest


class PrimesWithMemoTest(unittest.TestCase):
    def test_true(self):
        self.assertTrue(scramble('rkqodlw', 'world'))
        self.assertTrue(scramble('cedewaraaossoqqyt', 'codewars'))
        self.assertTrue(scramble('scriptjava', 'javascript'))
        self.assertTrue(scramble('scriptingjava', 'javascript'))

    def test_false(self):
        self.assertFalse(scramble('katas', 'steak'))

    def test_performance(self):
        level = 5
        s1 = ""
        for s in "zqwertyuiopasdfghjklxcvbnm": s1 += s * int(10 ** level)
        s1 += "z" * int(10 ** level * 2)
        s2 = ""
        for s in "zqwertyuiopasdfghjklxcvbnm": s2 += s * int(10 ** level)
        s2 += "z" * int(10 ** level * 2)
        actual = scramble(s1, s2)
        self.assertTrue(actual)


if __name__ == '__main__':
    unittest.main()

File: DotNet/Kyu5/Scramblies.cs
// https://www.codewars.com/kata/55c04b4cc56a697bb0000048/train/csharp

using System;
using System.Linq;

namespace DotNet.Kyu5;

public class Scramblies
{
    /// <summary>
    /// O(m * n)
    /// </summary>
    /// <param name="letterPools"></param>
    /// <param name="word"></param>
    /// <returns></returns>
    public static bool Scramble(string letterPools, string word)
    {
        var wordLetters = word.ToList();
        var letterPoolsLetters = letterPools.ToList();

        for (int wordIndex = 0; wordIndex < wordLetters.Count;)
        {
            var wordLetter = wordLetters[wordIndex];
            var hasLetter = false;

            for (int poolIndex = 0; poolIndex < letterPoolsLetters.Count;)
            {
                var poolLetter = letterPoolsLetters[poolIndex];
                if (wordLetter == poolLetter)
                {
                    wordLetters.RemoveAt(wordIndex);
                    letterPoolsLetters.RemoveAt(poolIndex);
                    hasLetter = true;
                    break;
                }

                poolIndex++;
            }

            if (!hasLetter) return false;
        }
        
        return wordLetters.Count == 0;
    }

    public static void Main()
    {
        Console.WriteLine($"{Scramble("rkqodlw","world")} == True");
        Console.WriteLine($"{Scramble("cedewaraaossoqqyt","codewars")} == True");
        Console.WriteLine($"{Scramble("katas","steak")} === False");
        Console.WriteLine($"{Scramble("scriptjavx","javascript")} == False");
        Console.WriteLine($"{Scramble("scriptsjava","javascripts")} == True");
        Console.WriteLine($"{Scramble("javscripts","javscripts")} == True");
        Console.WriteLine($"{Scramble("aabbcamaomsccdd","commas")} == True");
        Console.WriteLine($"{Scramble("commas","commas")} == True");
        Console.WriteLine($"{Scramble("sammoc","commas")} == True");

    }
}